### DISCUSSION TITLE
Wrong file icon for custom filetype

### DISCUSSION THREAD

**Discussion by nowaylifer** [ibhagwan/fzf-lua#1497] [Q&A] [answered]
I use 'mini.icons' and when modify filetype config as in the following example, the corresponding file's icon displayed incorrectly in fzf-lua window. Is there some workaround maybe?
 
```lua
vim.filetype.add({
  filename = {
    ["somefile"] = "lua",
  },
})
```

![Screenshot 2024-10-18 162654](https://github.com/user-attachments/assets/9548b9a4-a48e-428c-965e-cf3c1c9db799)

**Comment by ibhagwan** [comment]
> I use 'mini.icons' and when modify filetype config as in the following example, the corresponding file's icon displayed incorrectly in fzf-lua window. Is there some workaround maybe?

For performance reasons fzf-lua caches the internal mappings in mini, can you try loading fzf-lua for the first time only after calling `vim.filetype.add`?

**Reply by nowaylifer** [reply]
> > I use 'mini.icons' and when modify filetype config as in the following example, the corresponding file's icon displayed incorrectly in fzf-lua window. Is there some workaround maybe?
> 
> For performance reasons fzf-lua caches the internal mappings in mini, can you try loading fzf-lua for the first time only after calling `vim.filetype.add`?

Yes, I'm already lazy loading it with lazy.nvim

**Reply by ibhagwan** [reply]
> > > I use 'mini.icons' and when modify filetype config as in the following example, the corresponding file's icon displayed incorrectly in fzf-lua window. Is there some workaround maybe?
> > 
> > 
> > For performance reasons fzf-lua caches the internal mappings in mini, can you try loading fzf-lua for the first time only after calling `vim.filetype.add`?
> 
> Yes, I'm already lazy loading it with lazy.nvim

Alright I’ll check why this happens when I come back from my travel

#1498

**Comment by dpetka2001** [comment]
I notice something similar although not exactly the same. I have created a utility function to use for file browser-like experience (in an effort to get rid of other file explorer plugins, if it suits my needs after some time). The function is the following
```lua
function M.fzf_dirs(opts)
  local fzf = require("fzf-lua")
  local path = require("fzf-lua.path")
  -- require("fzf-lua.devicons").plugin_load("mini")
  local uv = vim.uv or vim.loop
  local cmd = "fd -d1"

  local function get_full_path(selected)
    if #selected < 1 then
      return
    end
    local entry = path.entry_to_file(selected[1], { cwd = uv.cwd() })
    if entry.path == "<none>" then
      return
    end
    -- Taken from require('fzf-lua').files, maybe there's a better way?
    local fullpath = entry.path or entry.uri and entry.uri:match("^%a+://(.*)")
    if not path.is_absolute(fullpath) then
      fullpath = path.join({ uv.cwd(), fullpath })
    end
    return fullpath
  end

  opts = opts or {}
  opts = {
    defaults = {},
    prompt = "Fzf Browser> " .. vim.fn.fnamemodify(uv.cwd(), ":~") .. "/",
    cwd = uv.cwd(),
    cwd_prompt_shorten_len = 32,
    cwd_prompt_shorten_val = 1,
    file_icons = true,
    fzf_opts = {
      ["--tiebreak"] = "end",
      ["--preview"] = {
        type = "cmd",
        fn = function(selected)
          local fullpath = get_full_path(selected)
          return string.format("command ls -Alhv --group-directories-first %s", fullpath)
        end,
      },
    },
    fn_transform = function(x)
      return fzf.make_entry.file(x, { file_icons = true, color_icons = true, cwd = uv.cwd() })
    end,
    actions = {
      ["default"] = {
        function(selected)
          local fullpath = get_full_path(selected)
          vim.cmd("cd " .. fullpath)
          -- fzf_lua.fzf_exec("fd -td", opts)
          opts.cwd = fullpath
          M.fzf_dirs()
        end,
      },
      ["alt-f"] = function(selected)
        local fullpath = get_full_path(selected)
        if uv.fs_stat(fullpath).type ~= "directory" then
          fullpath = vim.fn.fnamemodify(fullpath, ":p:h") .. "/"
        end
        vim.ui.input({ prompt = "File Name: " }, function(name)
          if name == nil then
            return
          end
          vim.cmd("e " .. fullpath .. name)
          vim.cmd("w ++p") -- will create directory if not exists, so you can write new/dir/some_file.lua
        end)
      end,
      ["alt-g"] = {
        function(selected)
          local cwd = selected[1] --[[@as string]]
          local parent = require("fzf-lua.path").parent(cwd, false)
          vim.cmd("cd ..")
          opts.cwd = parent
          M.fzf_dirs()
        end,
      },
    },
  }
  fzf.fzf_exec(cmd, opts)
end
```
I use LazyVim distro for my Neovim setup. I've observed the following: 
- When I press `<leader>fc` to invoke my custom function then I get no filetype icons. I get icons for `init.lua` and `README.MD and directories icons but not for simple files
- If I first open `find_files` then I have icons for everything. If I invoke my function and navigate to a folder that has Lua files, then I get no icons. When I open `find_files` again there are no icons this time, although the first time I opened it there were icons. When I open `buffers` I get no icons as well.
- If I first open `buffers` then I have icons. I open my custom function and it also has icons. I open `find_files` and it also has icons. 

From scourging through the code (although I'm not comfortable with the codebase and struggle a lot understanding), my guess is that `buffers` provider has a `make_entry.file` in which it doesn't load icons yet and later on constructs the entry with the `buficon`. 

So, if I open `buffers` first it caches the icons for the entry and then everything else also works normally? But if I open my custom function which also has a `make_entry.file` (with `file_icons = true` in contrast with what happens in `buffers` provider), then this for some reason doesn't load the icons and that stays cached and also affects the other pickers as well? 

LazyVim also uses `mini.icons` by default and I disabled them and enabled `nvim-web-devicons` in my personal configuration and can confirm that it doesn't happen then. It only happens with `mini.icons`.

Am I maybe not using the `make_entry.file` correctly in my custom function? If so, I'd appreciate some pointer as to how I can overcome this. For now I just first open `buffers`, so that everything works fine afterwards. Or if something else is amiss with `mini.icons` with regards to `fzf-lua`, I'll wait for an appropriate solution.

PS: If I do `=require("fzf-lua.devicons").state()` then `lua` extension has the `default_icon`. Actually, there are 2 entries, one with the correct icon and one with the default icon. I'm guessing the entry with the default icon comes from my custom function and it appears before the entry with the correct icon, so that's why it returns the default icon instead of the normal one??

**Reply by ibhagwan** [reply]
@dpetka2001, see my comment in https://github.com/ibhagwan/fzf-lua/issues/1498#issuecomment-2431219100, your code runs on the main thread therefore when the icons are queried they get cached which then proxies to the external process as well, that's why the ordering of what happens first matters, if the filename is cached first externally (without the custom filetypes) then the icon will not appear, if you first run your code which uses `fn_transform` on the main thread, it will cache the custom icon properly.

As I mentioned in my comment, due to the way mini.icons work and neovim not providing access to list custom filetypes I don't think I can solve this ATM.

**Reply by dpetka2001** [reply]
@ibhagwan No problem. Thank you for taking time to explain what happens. Launching buffers first and then my custom function is not a big deal for me.

**Reply by dpetka2001** [reply]
Hey @ibhagwan sorry for bothering you. I was wondering if you could lend a hand. 

I would like to add descriptions for the custom actions in the above picker. I know I should use `require("fzf-lua.config").set_action_helpstr` but having trouble doing it correctly. The above picker is not in my main `fzf-lua` configuration, but rather in a file under `/lua/util/fzf.lua` and I'm calling it via a keymap 
```lua
keys = {
      {
        "<leader>fc",
        Util.fzf.fzf_dirs,
        desc = "Fzf Browser",
      },
}
```
I would appreciate any help whenever you find some free time to do so.

**Reply by dpetka2001** [reply]
Ok I just did in the above snippet right after the `opts` table `config.set_action_helpstr(opts.actions["alt-f"], "create_file")` and it works for this. 

But it doesn't work for the `default` and `alt-g` action which are tables, because the assertion expects to be a function. But if I make them functions and not tables then the descriptions are ok but when I invoke these actions the fzf-lua UI flashes and I remember I had read that to avoid this blinking in the UI we should surround the action functions in a table to use the already existing fzf-lua process. Is there any other workaround for this that I can't think of?

**Reply by ibhagwan** [reply]
> config.set_action_helpstr

No need to use this anymore.

> But it doesn't work for the default and alt-g action which are tables, because the assertion expects to be a function.

In the action definition table add `desc` for the F1 help or `header` for the fzf header line:
```lua
  ['enter'] = {
    fn = function(sel, opts) … end,
    desc = 'f1-desc', -- make sure this is vimdoc help format as F1 ft is `help`
    header = 'fzf header',
    -- If the action reloads the picker add:
    -- reload = true,
    -- If the action performs a background task (use either this or reload, not both):
    -- exec_silent = true,
  }
```

**Reply by dpetka2001** [reply]
Thank you so much for this information. Really appreciate it. 

I tried `reload = true`, but I get a warning `actions with "reload" are only supported with string commands, using resume fallback`. But the command I use in `fzf_exec` is a string, it's `cmd = "fd -d1"`. But looking at the code [here](https://github.com/ibhagwan/fzf-lua/blob/ce1e24f2a48089b65deefcb0a71a9856839002b5/lua/fzf-lua/core.lua#L1067) I see that the function takes a `reload_cmd` parameter and when it gets called [here](https://github.com/ibhagwan/fzf-lua/blob/ce1e24f2a48089b65deefcb0a71a9856839002b5/lua/fzf-lua/core.lua#L437) it uses `opts.__reload_cmd or contents` to set the `reload_cmd`, which in my case should be `cmd = "fd -d1"` as the first parameter in `fzf_exec` that I use. So, why am I getting that warning?

I can totally get without `reload = true`, I'm just trying to understand why this is the case.

**Reply by ibhagwan** [reply]
> I can totally get without reload = true, I'm just trying to understand why this is the case.

You’re using a transform maybe? If so a different strategy is required due to the external process fzf-lua uses.

If you just wanna solve it usu can use `config.set_action_helpstr(<funcfion ptr>)` and use the function pointer from the `.fn` property from the action table definition you wrote.

**Reply by dpetka2001** [reply]
Yes, in the `opts` i use 
```lua
    fn_transform = function(x)
      return fzf.make_entry.file(x, { file_icons = true, color_icons = true, cwd = uv.cwd() })
    end,
```
to add icons and color to the entry. Does this affect the `reload = true` and changes the `reload_cmd` from the my own `cmd = "fd -d1"`?

Also thank you for the info about using the function pointer from `.fn` in the action table. That worked too. Good to know different ways of doing this.

**Reply by ibhagwan** [reply]
> Yes, in the `opts` i use
> 
> ```lua
>     fn_transform = function(x)
>       return fzf.make_entry.file(x, { file_icons = true, color_icons = true, cwd = uv.cwd() })
>     end,
> ```
> 
> to add icons and color to the entry. Does this affect the `reload = true` and changes the `reload_cmd` from the my own `cmd = "fd -d1"`?
> 
> Also thank you for the info about using the function pointer from `.fn` in the action table. That worked too. Good to know different ways of doing this.

This makes it a bit more difficult to use the reload action natively in fzf, so, as the warning informs, the process will restart instead, there is a way to make it work but it’s a bit more involved, you can take a look at the code of buffers/tabs/etc or git_stash but it would be a bit harder to understand why things are done this way unless you’re very familiar with the code base:
https://github.com/ibhagwan/fzf-lua/blob/ce1e24f2a48089b65deefcb0a71a9856839002b5/lua/fzf-lua/providers/git.lua#L234-L283

**Reply by dpetka2001** [reply]
No, I'm not that familiar with the code base at all. Like I already said I can totally do without `reload` as the table action handler works pretty well for my case.

By the way I also tried the `header` option like you said, but didn't work for me, so I just defined the headers in `fzf_opts["--header"]` instead and works perfectly fine.

Again, really appreciate you taking the time to interact with a user and help him clarify some things. Thank you!!

**Reply by ibhagwan** [reply]
> By the way I also tried the header option like you said, but didn't work for me, so I just defined the headers in fzf_opts["--header"] instead and works perfectly fine.

It should work, lmk if you want I can take a look at your code and see what’s wrong with the opt.

**Reply by dpetka2001** [reply]
Sure, this is the code of the custom function at the moment
```lua
function M.fzf_dirs(opts)
  local fzf = require("fzf-lua")
  local config = require("fzf-lua.config")
  local path = require("fzf-lua.path")
  local uv = vim.uv or vim.loop
  local utils = require("fzf-lua.utils")
  local cmd = "fd -d1"

  local function hl_validate(hl)
    return not utils.is_hl_cleared(hl) and hl or nil
  end

  local function ansi_from_hl(hl, s)
    return utils.ansi_from_hl(hl_validate(hl), s)
  end

  local function get_full_path(selected)
    if #selected < 1 then
      return
    end
    local entry = path.entry_to_file(selected[1], { cwd = uv.cwd() })
    if entry.path == "<none>" then
      return
    end
    -- Taken from require('fzf-lua').files, maybe there's a better way?
    local fullpath = entry.path or entry.uri and entry.uri:match("^%a+://(.*)")
    if not path.is_absolute(fullpath) then
      fullpath = path.join({ uv.cwd(), fullpath })
    end
    return fullpath
  end

  opts = opts or {}
  opts = {
    defaults = {},
    prompt = "Fzf Browser> " .. vim.fn.fnamemodify(uv.cwd(), ":~") .. "/",
    cwd = uv.cwd(),
    cwd_prompt_shorten_len = 32,
    cwd_prompt_shorten_val = 1,
    file_icons = true,
    fzf_opts = {
      ["--tiebreak"] = "end",
      ["--preview"] = {
        type = "cmd",
        fn = function(selected)
          local fullpath = get_full_path(selected)
          return string.format("command ls -Alhv --group-directories-first %s", fullpath)
        end,
      },
      -- ["--header"] = string.format(
      --   ":: <%s> to %s | <%s> to %s | <%s> to %s",
      --   ansi_from_hl("FzfLuaHeaderBind", "enter"),
      --   ansi_from_hl("FzfLuaHeaderText", "enter directory"),
      --   ansi_from_hl("FzfLuaHeaderBind", "alt-f"),
      --   ansi_from_hl("FzfLuaHeaderText", "create file"),
      --   ansi_from_hl("FzfLuaHeaderBind", "alt-g"),
      --   ansi_from_hl("FzfLuaHeaderText", "parent directory")
      -- ),
    },
    fn_transform = function(x)
      return fzf.make_entry.file(x, { file_icons = true, color_icons = true, cwd = uv.cwd() })
    end,
    actions = {
      ["enter"] = {
        fn = function(selected)
          local fullpath = get_full_path(selected)
          vim.cmd("tcd " .. fullpath)
          -- fzf_lua.fzf_exec("fd -td", opts)
          opts.cwd = fullpath
          M.fzf_dirs(opts)
        end,
        desc = "enter_directory",
        header = "test",
      },
      ["alt-f"] = function(selected)
        local fullpath = get_full_path(selected) --[[@as string]]
        if uv.fs_stat(fullpath).type ~= "directory" then
          fullpath = vim.fn.fnamemodify(fullpath, ":p:h") .. "/"
        end
        vim.ui.input({ prompt = "File Name: " }, function(name)
          if name == nil then
            return
          end
          vim.cmd("e " .. fullpath .. name)
          vim.cmd("w ++p") -- will create directory if not exists, so you can write new/dir/some_file.lua
        end)
      end,
      ["alt-g"] = {
        function(selected)
          local cwd = selected[1] --[[@as string]]
          local parent = require("fzf-lua.path").parent(cwd, false)
          vim.cmd("cd ..")
          opts.cwd = parent
          M.fzf_dirs(opts)
        end,
        desc = "parent_directory",
      },
    },
  }
  config.set_action_helpstr(opts.actions["alt-f"], "create_file")
  fzf.fzf_exec(cmd, opts)
end
```
and I just call it via a mapping with `lazy.nvim` keys as following
```lua
keys = {
      {
        "<leader>fc",
        Util.fzf.fzf_dirs,
        desc = "Fzf Browser",
      },
}
```

**Reply by ibhagwan** [reply]
Here you go @dpetka2001, slightly better UX using fzf's native reload so the process doesn't restart when you `enter|alt-g`, this also fixes the  header issue as  I forgot the headers are only activated when calling `  opts = core.set_header(opts, { "actions", "cwd" })`:

```lua
function M.fzf_dirs(opts)
  local fzf = require("fzf-lua")
  local core = fzf.core
  local path = fzf.path
  local utils = fzf.utils
  local config = fzf.config
  local shell = require("fzf-lua.shell")
  local uv = vim.uv or vim.loop

  local function hl_validate(hl)
    return not utils.is_hl_cleared(hl) and hl or nil
  end

  local function ansi_from_hl(hl, s)
    return utils.ansi_from_hl(hl_validate(hl), s)
  end

  local function get_full_path(selected)
    if #selected < 1 then
      return
    end
    local entry = path.entry_to_file(selected[1], { cwd = uv.cwd() })
    if entry.path == "<none>" then
      return
    end
    -- Taken from require('fzf-lua').files, maybe there's a better way?
    local fullpath = entry.path or entry.uri and entry.uri:match("^%a+://(.*)")
    if not path.is_absolute(fullpath) then
      fullpath = path.join({ uv.cwd(), fullpath })
    end
    return fullpath
  end

  opts = config.normalize_opts({
    defaults = {},
    cmd = "fd -d1",
    prompt = "Fzf Browser> " .. vim.fn.fnamemodify(uv.cwd(), ":~") .. "/",
    cwd = uv.cwd(),
    cwd_prompt_shorten_len = 32,
    cwd_prompt_shorten_val = 1,
    file_icons = true,
    fzf_opts = {
      ["--tiebreak"] = "end",
      ["--preview"] = {
        type = "cmd",
        fn = function(selected)
          local fullpath = get_full_path(selected)
          return string.format("command ls -Alhv --group-directories-first %s", fullpath)
        end,
      },
      -- ["--header"] = string.format(
      --   ":: <%s> to %s | <%s> to %s | <%s> to %s",
      --   ansi_from_hl("FzfLuaHeaderBind", "enter"),
      --   ansi_from_hl("FzfLuaHeaderText", "enter directory"),
      --   ansi_from_hl("FzfLuaHeaderBind", "alt-f"),
      --   ansi_from_hl("FzfLuaHeaderText", "create file"),
      --   ansi_from_hl("FzfLuaHeaderBind", "alt-g"),
      --   ansi_from_hl("FzfLuaHeaderText", "parent directory")
      -- ),
    },
    __fn_transform = function(x)
      return fzf.make_entry.file(x, { file_icons = true, color_icons = true, cwd = uv.cwd() })
    end,
    __fn_reload = function(_)
      return opts.cmd
    end,
    actions = {
      ["enter"] = {
        fn = function(selected)
          local fullpath = get_full_path(selected)
          vim.cmd("tcd " .. fullpath)
          -- fzf_lua.fzf_exec("fd -td", opts)
          opts.cwd = fullpath
          M.fzf_dirs(opts)
        end,
        desc = "enter_directory",
        header = "CD",
        reload = true,
      },
      ["alt-f"] = {
        fn = function(selected)
          local fullpath = get_full_path(selected) --[[@as string]]
          if uv.fs_stat(fullpath).type ~= "directory" then
            fullpath = vim.fn.fnamemodify(fullpath, ":p:h") .. "/"
          end
          vim.ui.input({ prompt = "File Name: " }, function(name)
            if name == nil then
              return
            end
            vim.cmd("e " .. fullpath .. name)
            vim.cmd("w ++p") -- will create directory if not exists, so you can write new/dir/some_file.lua
          end)
        end,
        desc = "create_file",
      },
      ["alt-g"] = {
        fn = function(selected)
          local cwd = selected[1] --[[@as string]]
          local parent = require("fzf-lua.path").parent(cwd, false)
          vim.cmd("cd ..")
          opts.cwd = parent
          M.fzf_dirs(opts)
        end,
        desc = "parent_directory",
        header = "Goto Parent",
        reload = true,
      },
    },
  }, {})

  -- build the "reload" cmd and remove '-- {+}' from the initial cmd
  local reload, id = shell.reload_action_cmd(opts, "{+}")
  local contents = reload:gsub("%-%-%s+{%+}$", "")
  opts.__reload_cmd = reload

  opts._fn_pre_fzf = function()
    shell.set_protected(id)
  end

  opts = core.set_header(opts, { "actions", "cwd" })
  return fzf.fzf_exec(contents, opts)
end
```

**Reply by dpetka2001** [reply]
Thank you very much. I'll definitely try this out and copy it in the util file. Not sure if I completely switch to this, because I'd prefer I understand what's going on. I have to improve my Lua knowledge and also get into FzfLua's codebase more in depth (at least in the amount possible by my current knowledge in programming, since I just started as a hobbyist at the start of this year). 

Again, really appreciate you taking time to interact with me. You've always been helpful and provided insight even in past issues/discussions that I created. Thank you!!!

**Reply by ibhagwan** [reply]
> Again, really appreciate you taking time to interact with me. You've always been helpful and provided insight even in past issues/discussions that I created. Thank you!!!

Ty for the kinds words @dpetka2001!

**Comment by ibhagwan** [comment, marked as answer]
https://github.com/ibhagwan/fzf-lua/issues/1498#issuecomment-2431815736
